### DISCUSSION TITLE
Nullable Reference types are really annoying to work with in expression trees.

### DISCUSSION THREAD

**Discussion by bbarry** [dotnet/csharplang#3387] [Language Ideas]
Perhaps as an alternative to #158, could the specification around flow analysis be changed to explicitly NOT support analysis inside expression tree types?

Nullable reference types make the missing null coalescing operators really stand out.

trying to use FluentValidation on type trees with nrt enabled:

    public class MyComplexTypeValidator : AbstractValidator<MyComplexType>
    {
        public MyComplexTypeValidator()
        {
            When(
                p => p.Db?.Table != null, //Func<MyComplexType, bool>
                () =>
                {
                    RuleFor(
    #pragma warning disable CS8602 // Dereference of a possibly null reference.
                        p => p.Db.Table //Expression<Func<MyComplexType, T>>
    #pragma warning restore CS8602 // Dereference of a possibly null reference.
                    ).ForEach(...)
                }
            );
        }
    }

**Comment by svick** [comment]
Why isn't [the null-forgiving operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) sufficient? It seems to work inside an expression tree. With that, your code would be:

```c#
RuleFor(
    p => p.Db!.Table //Expression<Func<MyComplexType, T>>
)
```
